### PR TITLE
Chaos selection history

### PR DESCRIPTION
**Pull request by jschueller**
Adds new methods to retrieve selection history from a FCE result
```
result = algo.getResult()

print(result.getIndicesHistory())
print(result.getCoefficientsHistory())
graph = result.drawSelectionHistory()
```

lars

notes:
- sofiane script does not account that LARS eventually works in a local selection of the basis
- might want to add Q2 history too
- only returns the last selection history & marginal

see also:
https://gist.github.com/sofianehaddad/0011686ed5eb917f1201383ee5069cbf
https://gist.github.com/sofianehaddad/a6f94bf992f8bde3810f9013bb1f3665

### CODE REVIEW

File: python/test/t_FunctionalChaos_mixed.py [hunk 1]
```diff
@@ -37,3 +37,12 @@
 result = algo.getResult()
 print(result.getRelativeErrors())
 assert result.getRelativeErrors()[0] < 1e-10
+
+# selection history
+indices = result.getIndicesHistory()
+coefs = result.getCoefficientsHistory()
+assert indices.getSize() > 0, "no indices slection"
```
Comment (mbaudin47) [inline python/test/t_FunctionalChaos_mixed.py:44]:
```suggestion
assert indices.getSize() > 0, "no indices selection"
```

**Review by mbaudin47 — CHANGES_REQUESTED** [review]

**Comment by mbaudin47** [comment]
A suggestion: If the output dimension is greater than 1, then an exception can be generated.

**Comment by jschueller** [comment]
TODO: new class to store results

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx [hunk 2]
```diff
@@ -339,6 +339,11 @@ void FunctionalChaosAlgorithm::run()
   }
   // Build the result
   result_ = FunctionalChaosResult(inputSample_, outputSample_, distribution_, transformation_, inverseTransformation_, basis, I_k, alpha_k, Psi_k, residuals, relativeErrors);
+
+  // set selection history
+  Collection<Point> coefficientsHistory;
+  Collection<Indices> indicesHistory(projectionStrategy_.getImplementation()->getSelectionHistory(coefficientsHistory));
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosAlgorithm.cxx:345]: I cannot understand how this works: ProjectionStrategyImplementation::getSelectionHistory() should throw an exception isn'it it?

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx [hunk 3]
```diff
@@ -300,6 +301,63 @@ void FunctionalChaosResult::load(Advocate & adv)
...
+Graph FunctionalChaosResult::drawSelectionHistory() const
+{
+  if (metaModel_.getOutputDimension() > 1)
+    throw NotYetImplementedException(HERE) << "drawSelectionHistory is only available for 1-d output dimension";
+  const UnsignedInteger size = indicesHistory_.getSize();
+  if (!size)
+    throw InvalidArgumentException(HERE) << "No selection history available";
+
+  // compute union of basis terms
+  std::map <UnsignedInteger, UnsignedInteger> indicesMap;
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx:333]: Do you need a map here? Or is an unordered_map ok? It is more efficient than a map.

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx [hunk 4]
```diff
@@ -202,6 +202,12 @@ void IntegrationStrategy::load(Advocate & adv)
   ProjectionStrategyImplementation::load(adv);
 }
 
+/* Selection history accessor */
+Collection<Indices> IntegrationStrategy::getSelectionHistory(Collection<Point> & coefficientsHistory) const
+{
+  coefficientsHistory = Collection<Point>();
```
Comment (regislebrun) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx:208]: Coupled with CleaningStrategy, an IntegrationStrategy may produce a selection history. It can be postpone to another PR of course.

**Review by regislebrun — APPROVED** [review]
Only minor comments

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx (hunk 4 above)
Comment (jschueller) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx:208]: yes, also we can do cleaning+lars ! :)

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx (hunk 3 above)
Comment (jschueller) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx:333]: unordered_map is ok too

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx [hunk 5]
```diff
@@ -300,6 +302,63 @@ void FunctionalChaosResult::load(Advocate & adv)
...
+{
+  if (metaModel_.getOutputDimension() > 1)
+    throw NotYetImplementedException(HERE) << "drawSelectionHistory is only available for 1-d output dimension";
+  const UnsignedInteger size = indicesHistory_.getSize();
+  if (!size)
+    throw InvalidArgumentException(HERE) << "No selection history available";
+
+  // compute union of basis terms
+  std::unordered_map <UnsignedInteger, UnsignedInteger> indicesMap;
+  UnsignedInteger count = 0;
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx:335]: What does this count?

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx [hunk 6]
```diff
@@ -300,6 +302,63 @@ void FunctionalChaosResult::load(Advocate & adv)
...
+    valuesX(i, 0) = i;
+  for (UnsignedInteger i = 0; i < size; ++ i)
+    for (UnsignedInteger j = 0; j < indicesHistory_[i].getSize(); ++ j)
+      valuesY(i + 1, indicesMap[indicesHistory_[i][j]]) = coefficientsHistory_[i][j];
+  Graph result("Selection history", "iteration", "coefficient", true, "topright");
+  for (UnsignedInteger i = 0; i < count; ++ i)
+  {
+    Curve curve(valuesX, valuesY.getMarginal(i));
+    curve.setColor(colors[i]);
+    curve.setLegend(OSS() << "term #" << uniqueBasisIndices[i]);
```
Comment (mbaudin47) [inline lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/FunctionalChaosResult.cxx:358]: I suggest to use the word "coefficient", because this is the way we describe each factor of a given basis function.
```suggestion
    curve.setLegend(OSS() << "Coef. #" << uniqueBasisIndices[i]);
```

File: lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx (hunk 4 above)
Comment (mbaudin47) [reply lib/src/Uncertainty/Algorithm/MetaModel/FunctionalChaos/IntegrationStrategy.cxx:208]: So is for the `LinearModelStepwiseAlgorithm`, which can be used for a PCE too.

File: python/src/FunctionalChaosResult_doc.i.in [hunk 7]
```diff
@@ -160,3 +160,51 @@ Returns
...
+
+%feature("docstring") OT::FunctionalChaosResult::getCoefficientsHistory
+"The coefficients values selection history accessor.
+
+This is only availble with LARS, and when the output dimension is 1.
+
+Returns
+-------
+coefficientsHistory : 2-d sequence of float
+    The coefficients values selection history"
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:186]: It seems to me that the number of rows and columns should be presented. A pointer to the theory help page `polynomial_sparse_least_squares` will make this easier to understand.

File: python/src/FunctionalChaosResult_doc.i.in [hunk 8]
```diff
@@ -160,3 +160,51 @@ Returns
...
+
+%feature("docstring") OT::FunctionalChaosResult::getIndicesHistory
+"The basis indices selection history accessor.
+
+This is only availble with LARS, and when the output dimension is 1.
+
+Returns
+-------
+indicesHistory : 2-d sequence of int
+    The basis indices selection history"
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:198]: Idem.

File: python/src/FunctionalChaosResult_doc.i.in [hunk 9]
```diff
@@ -160,3 +160,51 @@ Returns
...
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::FunctionalChaosResult::setSelectionHistory
+"The basis coefficients and indices accessor.
+
+Parameters
+----------
+indicesHistory : 2-d sequence of int
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:207]: Idem. Furthermore, the two input arguments must be consistent.

File: python/src/FunctionalChaosResult_doc.i.in [hunk 10]
```diff
@@ -160,3 +160,51 @@ Returns
 -------
 transformation : :class:`~openturns.Function`
     Transformation :math:`T` such that :math:`T(\vect{X}) = \vect{Z}`."
+    
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::FunctionalChaosResult::drawSelectionHistory
+"Draw the basis selection history.
+
+This is only availble with LARS, and when the output dimension is 1.
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:169]:
```suggestion
This is only availble with :class:`~openturns.LARS`, and when the output dimension is 1.
```

File: python/src/FunctionalChaosResult_doc.i.in [hunk 11]
```diff
@@ -160,3 +160,51 @@ Returns
...
+-------
+graph : :class:`~openturns.Graph`
+    The evolution of the basis coefficients at each selection iteration"
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::FunctionalChaosResult::getCoefficientsHistory
+"The coefficients values selection history accessor.
+
+This is only availble with LARS, and when the output dimension is 1.
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:181]:
```suggestion
This is only availble with :class:`~openturns.LARS`, and when the output dimension is 1.
```

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
Thank you very much for this long-awaited feature! I feel that there is room for improvement. A short example would make the feature more appreciated. The important point here is to show the plot. This might be done e.g. within the API help page of `LARS` as a plot_LARS.py script, by pasting the code you used for this very PR (no extra work is needed here). Forward and backward pointers to the `LARS` class will be welcome too. The figure you created could be added to the `polynomial_sparse_least_squares` theory help page at no extra cost, with a clear advantage when the review of OT1.22 features will be done in a few months: the picture will be ready to go.

File: python/src/FunctionalChaosResult_doc.i.in [hunk 12]
```diff
@@ -160,3 +160,51 @@ Returns
...
+-------
+coefficientsHistory : 2-d sequence of float
+    The coefficients values selection history"
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::FunctionalChaosResult::getIndicesHistory
+"The basis indices selection history accessor.
+
+This is only availble with LARS, and when the output dimension is 1.
```
Comment (mbaudin47) [inline python/src/FunctionalChaosResult_doc.i.in:193]:
```suggestion
This is only available with :class:`~openturns.LARS`, and when the output dimension is 1.
```

**Comment by mbaudin47** [comment]
I created a script at [gist](https://gist.github.com/mbaudin47/26eadbd1a503c51326ac36abc89eeb59) to experiment with the PR.

Sometimes a curve can go down as in the following example.

image

One of the most interesting pictures is from GSobol with a = [0, 1].

image

**Comment by jschueller** [comment]
lets merge it since everyone agreed
